Implement --unstable-presymbolicate

vvuk opened the pull request:
Adds `--unstable-presymbolicate`, which will output ` .syms.json` file alongside the regular profile output. If that file is present when the server is fired up (based on the profile name, i.e. `foo.json` -> `foo.syms.json`), it is parsed and the data is added to be used when symbolicating. If a library has presymbolicated info (by debugid), _only_ that info is used. There's no attempt to fall back to on-disk etc.

This allows a capture to be entirely self-contained, so that no symbol servers or on-disk data are needed for looking at the profile, which is very useful for running in CI or other locations where publishing the debuginfo isn't necessarily useful otherwise.

mstange commented on fxprof-processed-profile/src/global_lib_table.rs:
```diff
@@ -73,4 +94,26 @@ impl Serialize for GlobalLibIndex {
 /// The handle for a library, obtained from [`Profile::add_lib`](crate::Profile::add_lib).
 #[derive(Debug, Clone, Copy, PartialOrd, Ord, PartialEq, Eq, Hash)]
-pub struct LibraryHandle(usize);
+pub struct LibraryHandle(pub(crate) usize);
```
Is the `pub(crate)` still necessary?

mstange commented on fxprof-processed-profile/src/global_lib_table.rs:
```diff
@@ -14,6 +15,11 @@ pub struct GlobalLibTable {
+    /// library, so that we can potentially provide symbolication info ahead of time.
+    /// This is here instead of in `LibraryInfo` because we don't want to serialize it,
+    /// and because it's currently a hack.
+    all_libs_seen_rvas: Vec<Option<BTreeSet<u32>>>,
```
Please add a comment saying that the Vec is indexed by `LibraryHandle.0`.

Alternatively, you could make the Vec indexed by `GlobalLibIndex.0`, needing one less translation in `add_lib_used_rva` and at that point you could also remove the Option and always initialize it to the empty set.

vvuk replied:
Yeah, good idea. FWIW, `GlobalLibIndex` is a confusing name, maybe just `UsedLibraryHandle`?

mstange replied:
`UsedLibraryHandle` sounds good

mstange reviewed (commented):
Will review more tomorrow.

mstange commented on fxprof-processed-profile/src/global_lib_table.rs:
```diff
@@ -73,4 +94,26 @@ impl Serialize for GlobalLibIndex {
+    fn next(&mut self) -> Option<Self::Item> {
+        self.global_lib_table
+            .used_libs
+            .get(self.next_used_lib_index)
```
Could use a question mark here, so you don't need the `map`.

mstange commented on samply/src/shared/symbol_precog.rs:
```diff
@@ -0,0 +1,386 @@
+
+impl<'de> Deserialize<'de> for StringTableIndex {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        let value = String::deserialize(deserializer)?;
```
Hmm, why `String::deserialize`? Is it in quotes? It shouldn't be.

mstange commented on samply/src/shared/symbol_precog.rs:
```diff
@@ -0,0 +1,386 @@
+                        //eprintln!("lookup_sync: 0x{:x} -> {}", rva, info.symbol.0);
+                        return Some(wholesym::SyncAddressInfo {
+                            symbol: wholesym::SymbolInfo {
+                                address: rva,
+                                size: None,
```
I'd prefer it if you stored the correct symbol address. Sometimes you have many different addresses in the same function, and we want the front-end to create a shared symbol for those different addresses. Here you're giving them all different symbol addresses, so the front-end will think they're different symbols with the same name.

mstange commented on wholesym/src/config.rs:
```diff
@@ -20,6 +33,7 @@ pub struct SymbolManagerConfig {
     pub(crate) use_spotlight: bool,
     pub(crate) debuginfod_cache_dir_if_not_installed: Option<PathBuf>,
     pub(crate) debuginfod_servers: Vec<(String, PathBuf)>,
+    pub(crate) precog_data: Option<PrecogDataContainer>,
```
As I said in https://github.com/Unity-Technologies/usamply/pull/1#discussion_r1601910497, I would prefer storing it on the helper rather than in the config, and without an extra wrapper struct.

vvuk replied:
Oh, I misunderstood that comment! I missed the part about keeping it in the helper. Will do.

vvuk replied:
(The wrapper struct was to allow for Debug FWIW -- `SymbolMapTrait` does not imply Debug, so I would have had to define another trait and make sure all symbol map objects where `Debug`, or implement debug manually for `SymbolManagerConfig`)

mstange commented on wholesym/src/config.rs:
```diff
@@ -1,12 +1,25 @@
 /// The configuration of a [`SymbolManager`](crate::SymbolManager).
 ///
 /// Allows specifying various sources of symbol files.
-#[derive(Debug, Clone, Default)]
```
Please keep the config Clone.

vvuk replied:
Ah yes, this can come back. When I was using a trait object in here, I had all sorts of problems but since that's gone this should be fine. It's moving to the helper anyway.

vvuk commented:
Ok, implemented proper address/symbol info and properly implemented frames. Should have addressed all the review feedback here. I didn't make the `GlobalLibIndex` -> `UsedLibraryHandle` change, that can be a future thing
